README.md
![](logo_transparent.png)

# Millfork

A middle-level programming language targeting 6502-based, 8080-based, Z80-based and 6809-based microcomputers. 

For binary releases, see: [https://github.com/KarolS/millfork/releases](https://github.com/KarolS/millfork/releases)
(latest: 0.3.30).  
For build instructions, see [Build instructions](./COMPILING.md).

## Features

* high performance, due to being designed and optimized for 8-bit microprocessors

* multiple targets:

    * Commodore 64 (the primary target)
    
    * Commodore 64 with SuperCPU (experimental, incomplete and very buggy)
    
    * other Commodore computers: C16, Plus/4, C128, PET, VIC-20 (stock or with RAM extensions)
    
    * other 6502-based machines: Famicom/NES, Atari Lynx, Atari 8-bit computers, BBC Micro, Apple II+/IIe/Enhanced IIe, Atari 2600 (experimental), Commander X16 (experimental)
    
    * Z80-based machines: ZX Spectrum 48k, NEC PC-88, Amstrad CPC, MSX, TRS-80 Model 1 and 3, Robotron Z1013
    
    * CP/M
    
    * Game Boy (experimental)
    
    * Tandy Color Computer (experimental)
    
    * MS-DOS (very experimental, via 8080-to-8086 translation)

* multiple supported target processors:

    * well supported: MOS 6502, Ricoh 2A03/2A07, WDC 65C02, Intel 8080, Intel 8085, Zilog Z80
    
    * reasonably well supported: Sharp LR35902, CSG 65CE02, Motorola 6809
    
    * partially supported: Hudson Soft HuC6280, WDC 65816, Intel 8086

* inline assembly

* simple macros

* pay only for what you use: not a single byte of memory is used unless for code or explicitly declared variables

* a simple memory model that avoids using the stack

* multi-pass whole-program optimizer (that will even optimize your hand-written assembly if you ask it to)

* support for multi-file programs (Commodore only) and banked cartridges

## Licensing

The compiler is distributed under GPLv3 (see [LICENSE](LICENSE)).

The standard include files (located in the `include` directory) are distributed under a more permissive Zlib license (see [include/LICENSE](include/LICENSE)).
Therefore, no attribution is needed if you are developing and distributing Millfork programs.

The documentation is distributed under the [CC-0 license](https://creativecommons.org/publicdomain/zero/1.0/).

## More info

* [Documentation](docs/README.md) (external version: [https://millfork.readthedocs.io](https://millfork.readthedocs.io))

* [Example programs](examples/README.md)



CHANGELOG.md
# Change log

## 0.3.30 (2021-12-15)

* Added volatile structure fields (#112).

* Added `this.function` as the alias for the current function (#118).

* Added support for constant evaluation in `file` expressions (#114).

* Allowed declaring local constants and passing untyped parameters for macros.

* Allowed treating bare function name as a pointer to it.

* Added Mesen, ld65 and "raw" label file formats (#128).

* Commodore: the address used by SYS is now determined automatically instead of hardcoded (#111).

* C64: Fixed address for `sid_v1_sr` (#115).

* EasyFlash: Fixed address for `switch_hirom` (#121).

* GB: Fixed standard library (thanks to @retrac0).
  
* Commander X16: Updated platform definition file (thanks to @mookiexl).

* 65CE02: Full assembly support.

* R800: Full assembly support.

* Various miscompilation fixes (#123, #125) and parser fixes (e.g. #120).

* 6809: Various optimizations.

* Improvements related to constant evaluation.

## 0.3.28 (2021-05-24)

* Officially deprecated decimal operators with apostrophes.

* Added optimization hints.

* Added `utf32be`, `utf32le`, `cp1253`, `cp1254`, `cp1257`, `geos_de` encodings.

* Allowed for underscores in numeric literals for readability purposes, similar to several other programming languages. 

* Added a warning for comparisons between bytes and pointers (#110).

* Fixed escape sequences in many encodings.

* Fixed and documented absolute module imports (#106)
  
* Fixed and optimized sign extension.
  
* Fixed optimizations involving unused labels.
  
* Fixed pointer types to type aliases.

* Fixed parsing of Intel hex literals of the form `0BH`, `0B0H` etc. 

* 6809: Fixed flow analysis in optimization.

* Optimization of certain bitmask operations.

* Parsing optimizations.

## 0.3.26 (2021-03-01)

* Array fields in structs.

* Various Apple II-related improvements, including ProDOS support (thanks to @retrac0).

* Segment-related constants now match their equivalents from the platform definition. Missing constants have been defined.

* Constants with heap start and segment start are now generated properly.

* Signed multiplication support for `sbyte` and `signed16`.

* Heavily experimental `typeof` builtin.

* Self-modifying assembly code is now supported (#101).

* Successful compilation now prints result program size.

* Warning about data not being included in the output file.

* Warnings can now be enabled and disabled individually.

* Imported modules are now identified by their full relative path, not just the token used in the `import` statement (#89).

* 6502: Fixed sbyte to word promotions in certain contexts.

* 8080: Fixed compilation of sign extension of `sbyte` values into the BC register pair.

* Fixed negative constant folding.

* Fixed optimizations around macro invocations.

* 6502: Fixed code deduplication in presence of trampolined functions.

* Optimized word shifts for between 7 and 12 bits.

* Allowed new lines after `=`.

* Various optimization improvements.

* Improved some error messages (thanks to @agg23).

* Other fixes and improvements.

## 0.3.24 (2020-12-02)

* Preliminary support for TRS-80 Model 1 and 3 running TRS-DOS.

* Preliminary support for Robotron Z1013.

* Allowed defining entry points other than the start of the segment for Atari, ZX Spectrum, CoCo, Z1013 and TRS-80. (#78)

* Allowed the `:` operator in const-pure functions.

* Added `pointer.interrupt` and `pointer.kernal_interrupt` types.

* Implemented `readline` and `readword` for VIC-20.

* `init_rand_seed` uses the POKEY on Atari.

* Useless labels are no longer emitted into the label file.

* VIC-20: added `readline` and `readword`.

* Atari: use POKEY for randomness source.

* New output format elements: ASCII string, program name.

* Fix: Pointers to functions with parameters (#86)/

* Fix: more instances of memset loops should be detected and optimized (#59).

* Fix: things mentioned in the segment layout should not be deleted even if unused.

* Fix: `endaddr+N` output format.

* 65816: some code generation fixes.

* 8080: word negation now works.

* Various optimization improvements.

* Various other fixes.

* Improved some error messages.

* Even more new Atari examples (thanks to @zbyti).

* Build process slightly changed.

## 0.3.22 (2020-09-15)

* Added local labels in assembly.

* Added alternate decimal operators (with `$` instead of `'`).

* **Potentially breaking change!** Identifiers no longer can end with `$`.

* Added `z80next` as an alternate name for the ZX Spectrum Next's processor (#55).

* Added encodings: `brascii`, `macroman`, `dmcs`, `lics`.

* Improved some error messages.

* Fix: interrupt functions written in assembly no longer have the default prologue (#62).

* Fixed the `a8_os` module (#58).

* Fixed evaluation of division of large constants.

* Fix: Structure alignment is now respected for substructures.

* X16: Fixed the address of `vera_dc_hscale_hstop` register (#54) (thanks to @Kobrasadetin).

* Fixed evaluation of more complex boolean expressions (#56).


src/main/scala/millfork/Platform.scala
package millfork

import java.io.{File, StringReader}
import java.nio.charset.StandardCharsets
import java.nio.file.{Files, Paths}
import java.util.Locale

import millfork.error.Logger
import millfork.output._
import millfork.parser.{TextCodec, TextCodecRepository, TextCodecWithFlags}
import org.apache.commons.configuration2.INIConfiguration

/**
  * @author Karol Stasiak
  */

object OutputStyle extends Enumeration {
  val Single, PerBank, LUnix = Value
}

class Platform(
                val cpu: Cpu.Value,
                val flagOverrides: Map[CompilationFlag.Value, Boolean],
                val startingModules: List[String],
                val defaultCodec: TextCodec,
                val screenCodec: TextCodec,
                val features: Map[String, Long],
                val outputPackagers: Map[String, OutputPackager],
                val defaultOutputPackager: OutputPackager,
                val codeAllocators: Map[String, UpwardByteAllocator],
                val variableAllocators: Map[String, VariableAllocator],
                val zpRegisterSize: Int,
                val freeZpBytes: List[Int],
                val fileExtension: String,
                val generateBbcMicroInfFile: Boolean,
                val generateGameBoyChecksums: Boolean,
                val bankNumbers: Map[String, Int],
                val bankLayouts: Map[String, Seq[String]],
                val bankFill: Map[String, Int],
                val defaultCodeBank: String,
                val ramInitialValuesBank: Option[String],
                val outputLabelsFormat: DebugOutputFormat,
                val outputStyle: OutputStyle.Value
              ) {

  def hasZeroPage: Boolean = cpuFamily == CpuFamily.M6502

  def cpuFamily: CpuFamily.Value = CpuFamily.forType(this.cpu)

  def isBigEndian: Boolean = CpuFamily.isBigEndian(cpuFamily)

  private def bankStart(bank: String): Int = codeAllocators(bank).startAt min variableAllocators(bank).startAt

  private def bankEndBefore(bank: String): Int = codeAllocators(bank).endBefore max variableAllocators(bank).endBefore

  def isUnsafeToJump(bank1: String, bank2: String): Boolean = {
    if (bank1 == bank2) return false
    val s1 = bankStart(bank1)
    val s2 = bankStart(bank2)
    val e1 = bankEndBefore(bank1)
    val e2 = bankEndBefore(bank2)
    e2 > s1 && s2 < e1
  }

  lazy val banksExplicitlyWrittenToOutput: Set[String] = bankNumbers.keySet.filter(b => defaultOutputPackager.writes(b)).toSet

  def getMesenLabelCategory(bank: String, address: Int): Char = {
    // see: https://www.mesen.ca/docs/debugging/debuggerintegration.html#mesen-label-files-mlb
    val start = bankStart(bank)
    val end = bankEndBefore(bank)
    if (start > address || address >= end) 'G'
    else if (banksExplicitlyWrittenToOutput(bank)) 'P'
    else if (bank == "default") 'R'
    else 'W' // TODO: distinguish between W and S
  }
}

object Platform {

  def lookupPlatformFile(includePath: List[String], platformName: String, textCodecRepository: TextCodecRepository)(implicit log: Logger): Platform = {
    includePath.foreach { dir =>
      val file = Paths.get(dir, platformName + ".ini").toFile
      log.debug("Checking " + file)
      if (file.exists()) {
        return load(file, textCodecRepository)
      }
      val file2 = Paths.get(dir, "platform", platformName + ".ini").toFile
      log.debug("Checking " + file2)
      if (file2.exists()) {
        return load(file2, textCodecRepository)
      }
    }
    log.fatal(s"Platform definition `$platformName` not found", None)
  }

  def load(file: File, textCodecRepository: TextCodecRepository)(implicit log: Logger): Platform = {
    val conf = new INIConfiguration()
    val bytes = Files.readAllBytes(file.toPath)
    conf.read(new StringReader(new String(bytes, StandardCharsets.UTF_8)))

    val cs = conf.getSection("compilation")
    val cpu = Cpu.fromString(cs.get(classOf[String], "arch", "strict"))
    val value65816 = cs.get(classOf[String], "emit_65816", "")
    val flagOverrides = (value65816.toLowerCase match {
      case "" => Nil
      case "false" | "none" | "no" | "off" | "0" =>
        List(
          CompilationFlag.EmitEmulation65816Opcodes -> false,
          CompilationFlag.EmitNative65816Opcodes -> false,
          CompilationFlag.ReturnWordsViaAccumulator -> false)
      case "emulation" =>
        List(
          CompilationFlag.EmitEmulation65816Opcodes -> true,
          CompilationFlag.EmitNative65816Opcodes -> false,
          CompilationFlag.ReturnWordsViaAccumulator -> false)
      case "native" =>
        List(
          CompilationFlag.EmitEmulation65816Opcodes -> true,
          CompilationFlag.EmitNative65816Opcodes -> true)
      case _ =>
        log.error(s"Unsupported `emit_65816` value: $value65816")
        Nil
    }).toMap ++ CompilationFlag.fromString.flatMap { case (k, f) =>
      val value = cs.get(classOf[String], k, "")
      value.toLowerCase match {
        case "" | null => None
        case "false" | "off" | "no" | "0" => Some(f -> false)
        case "true" | "on" | "yes" | "1" => Some(f -> true)
        case _ =>
          log.error(s"Unsupported `$k` value: $value")
          None
      }
    }
    val startingModules = cs.get(classOf[String], "modules", "").split("[, ]+").filter(_.nonEmpty).toList
    val zpRegisterSize = cs.get(classOf[String], "zeropage_register", "").toLowerCase match {
      case "" | null => if (CpuFamily.forType(cpu) == CpuFamily.M6502) 4 else 0
      case "true" | "on" | "yes" => 4
      case "false" | "off" | "no" | "0" => 0
      case x => x.toInt
    }
    if (zpRegisterSize < 0 || zpRegisterSize > 128) {
      log.error("Invalid zeropage register size: " + zpRegisterSize)
    }

    val codecName = cs.get(classOf[String], "encoding", "ascii")
    val srcCodecName = cs.get(classOf[String], "screen_encoding", codecName)
    val TextCodecWithFlags(codec, czt, clp, _) = textCodecRepository.forName(codecName, None, log)
    if (czt) {
      log.error("Default encoding cannot be zero-terminated")
    }
    if (clp) {
      log.error("Default encoding cannot be length-prefixed")
    }
    if (codec.stringTerminator.length != 1) {
      log.warn("Default encoding should be byte-based")
    }
    val TextCodecWithFlags(srcCodec, szt, slp, _) = textCodecRepository.forName(srcCodecName, None, log)
    if (szt) {
      log.error("Default screen encoding cannot be zero-terminated")
    }
    if (slp) {
      log.error("Default screen encoding cannot be length-prefixed")
    }

    val as = conf.getSection("allocation")

    val banks = as.get(classOf[String], "segments", "default").split("[, ]+").filter(_.nonEmpty).toList
    if (!banks.contains("default")) {
      log.error("A segment named `default` is required")
    }
    if (banks.toSet.size != banks.length) {
      log.error("Duplicate segment name")
    }
    val BankRegex = """\A[A-Za-z0-9_]+\z""".r
    banks.foreach {
      case BankRegex(_*) => // ok
      case b => log.error(s"Invalid segment name: `$b`")
    }

    val bankStarts = banks.map(b => b -> (as.get(classOf[String], s"segment_${b}_start") match {
      case "" | null => log.error(s"Undefined segment_${b}_start"); 0
      case x => parseNumber(x)
    })).toMap
    val bankDataStarts = banks.map(b => b -> (as.get(classOf[String], s"segment_${b}_datastart", "after_code") match {
      case "" | "after_code" => None
      case x => Some(parseNumber(x))
    })).toMap
    val bankEnds = banks.map(b => b -> (as.get(classOf[String], s"segment_${b}_end") match {
      case "" | null => log.error(s"Undefined segment_${b}_end"); 0xffff
      case x => parseNumber(x)
    })).toMap
    val bankCodeEnds = banks.map(b => b -> (as.get(classOf[String], s"segment_${b}_codeend", "") match {
      case "" => bankEnds(b)
      case x => parseNumber(x)
    })).toMap
    val defaultCodeBank = as.get(classOf[String], "default_code_segment") match {
      case "" | null => "default"
      case x => x
    }
    val ramInitialValuesBank = as.get(classOf[String], "ram_init_segment") match {
      case "" | null => None
      case "default" =>
        log.error("Cannot use default as ram_init_segment")
        None
      case x if banks.contains(x) => Some(x)
      case x =>
        log.error("Invalid ram_init_segment: " + x)
        None
    }
    // used by 65816 and in NES debugging:
    val bankNumbers = banks.map(b => b -> (as.get(classOf[String], s"segment_${b}_bank", "00") match {
      case "" => 0
      case x => parseNumber(x)
    })).toMap
    val bankFills = banks.map(b => b -> (as.get(classOf[String], s"segment_${b}_fill", "00") match {
      case "" => 0
      case x => parseNumber(x)
    })).toMap
    // needed for ZX81
    val bankLayouts = banks.map(b => b -> {
      val layout = as.get(classOf[String], s"segment_${b}_layout", "main,*").split(',').map(_.trim).toSeq
      if (layout.isEmpty) {
        log.error(s"Layout for segment $b shouldn't be empty")
      } else {
        if (!layout.contains("*")) {
          log.error(s"Layout for segment $b should contain *")
        }
        if (layout.toSet.size != layout.size) {
          log.error(s"Layout for segment $b should contains duplicates")
        }
        if (ramInitialValuesBank.contains(b) && layout.last != "*") {
          log.warn(s"Layout for the ram_init_segment $b does not end with *")
        }
      }
      layout
    }).toMap

    // TODO: validate stuff
    banks.foreach(b => {
      if (bankNumbers(b) < 0 || bankNumbers(b) > 255) log.error(s"Segment $b has invalid bank")
      if (bankStarts(b) >= bankCodeEnds(b)) log.error(s"Segment $b has invalid range")
      if (bankCodeEnds(b) > bankEnds(b)) log.error(s"Segment $b has invalid range")
      if (bankStarts(b) >= bankEnds(b)) log.error(s"Segment $b has invalid range")
      bankDataStarts(b).foreach(dataStarts => if (dataStarts >= bankEnds(b)) log.error(s"Segment $b has invalid range"))
    })

    val freePointers: Option[List[Int]] = as.get(classOf[String], "zp_pointers", "") match {
      case "all" => Some(List.tabulate(128)(_ * 2))
      case "" => None
      case xs => Some(xs.split("[, ]+").flatMap(s => parseNumberOrRange(s, 2)).toList)
    }
    val freeExplicitBytes: Option[List[Int]] = as.get(classOf[String], "zp_bytes", "") match {
      case "all" => Some(List.tabulate(256)(identity))
      case "" => None
      case xs => Some(xs.split("[, ]+").flatMap(s => parseNumberOrRange(s, 1)).toList)
    }

    val freeZpBytes: List[Int] = (freePointers, freeExplicitBytes) match {
      case (Some(l), None) => l.flatMap(i => List(i, i+1))
      case (None, Some(l)) => l
      case (None, None) => List.tabulate(256)(identity)
      case (Some(_), Some(l)) =>
        log.error(s"Cannot define both zp_pointers and zp_bytes")
        l
    }

    val codeAllocators = banks.map(b => b -> new UpwardByteAllocator(bankStarts(b), bankCodeEnds(b) + 1))
    val variableAllocators = banks.map(b => b -> new VariableAllocator(
      if (b == "default" && CpuFamily.forType(cpu) == CpuFamily.M6502) freeZpBytes else Nil, bankDataStarts(b) match {
        case None => new AfterCodeByteAllocator(bankStarts(b), bankEnds(b) + 1)
        case Some(start) => new UpwardByteAllocator(start, bankEnds(b) + 1)
      }))

    val os = conf.getSection("output")
    def parseOutputPackager(s: String): OutputPackager = SequenceOutput(s.split("[, \n\t\r]+").filter(_.nonEmpty).map {
      case "startaddr" => StartAddressOutput(0)
      case "startaddr_be" => StartAddressOutputBe(0)
      case "startpage" => StartPageOutput
      case "endaddr" => EndAddressOutput(0)
      case "endaddr_be" => EndAddressOutputBe(0)
      case l if l.startsWith("\"") && l.endsWith("\"") => StringOutput(l.substring(1, l.length - 1))
      case l if l.startsWith("programname-") => ProgramNameOutput(parseNumber(l.stripPrefix("programname-")))
      case l if l.startsWith("startaddr+") => StartAddressOutput(parseNumber(l.stripPrefix("startaddr+")))
      case l if l.startsWith("startaddr-") => StartAddressOutput(-parseNumber(l.stripPrefix("startaddr-")))
      case l if l.startsWith("startaddr_be+") => StartAddressOutputBe(parseNumber(l.stripPrefix("startaddr_be+")))
      case l if l.startsWith("startaddr_be-") => StartAddressOutputBe(-parseNumber(l.stripPrefix("startaddr_be-")))
      case l if l.startsWith("endaddr+") => EndAddressOutput(parseNumber(l.stripPrefix("endaddr+")))
      case l if l.startsWith("endaddr-") => EndAddressOutput(-parseNumber(l.stripPrefix("endaddr-")))
      case l if l.startsWith("endaddr_be+") => EndAddressOutputBe(parseNumber(l.stripPrefix("endaddr_be+")))
      case l if l.startsWith("endaddr_be-") => EndAddressOutputBe(-parseNumber(l.stripPrefix("endaddr_be-")))
      case "pagecount" => PageCountOutput
      case "allocated" => AllocatedDataOutput
      case "length" => AllocatedDataLength(0)
      case "length_be" => AllocatedDataLengthBe(0)
      case l if l.startsWith("length+") => AllocatedDataLength(parseNumber(l.stripPrefix("length+")))
      case l if l.startsWith("length_be+") => AllocatedDataLengthBe(parseNumber(l.stripPrefix("length_be+")))
      case l if l.startsWith("length-") => AllocatedDataLength(-parseNumber(l.stripPrefix("length-")))
      case l if l.startsWith("length_be-") => AllocatedDataLengthBe(-parseNumber(l.stripPrefix("length_be-")))
      case "length_be" => AllocatedDataLengthBe(0)
      case "d88" => D88Output
      case "tap" => new TapOutput("main")
      case l if l.startsWith("tap:") => new TapOutput(l.stripPrefix("tap:").trim)
      case "trscmd" => new TrsCmdOutput("main")
      case l if l.startsWith("trscmd:") => new TrsCmdOutput(l.stripPrefix("trscmd:").trim)
      case n => n.split(":").filter(_.nonEmpty) match {
        case Array(b, s, e) => BankFragmentOutput(b, parseNumber(s), parseNumber(e))
        case Array(s, e) =>
          s match {
            case "addr" =>
              val (symbol, bonus) = parseSymbolAndBonus(e)
              SymbolAddressOutput(symbol, bonus)
            case "addr_be" =>
              val (symbol, bonus) = parseSymbolAndBonus(e)
              SymbolAddressOutputBe(symbol, bonus)
            case _ =>
              CurrentBankFragmentOutput(parseNumber(s), parseNumber(e))
          }
        case Array(b) => try {
          ConstOutput(parseNumber(b).toByte)
        } catch {
          case _:NumberFormatException => log.fatal(s"Invalid output format: `$b`")
        }
        case x => log.fatal(s"Invalid output format: `$x`")
      }
    }.toList)
    val outputPackagers = banks.flatMap{ b =>
      val f = os.get(classOf[String], "format_segment_" + b, null)
      if (f eq null) {
        None
      } else {
        Some(b -> parseOutputPackager(f))
      }
    }.toMap
    val defaultOutputPackager = parseOutputPackager(os.get(classOf[String], "format", ""))
    val fileExtension = os.get(classOf[String], "extension", ".bin")
    val generateBbcMicroInfFile = os.get(classOf[Boolean], "bbc_inf", false)
    val generateGameBoyChecksums = os.get(classOf[Boolean], "gb_checksum", false)
    val outputStyle = os.get(classOf[String], "style", "single") match {
      case "" | "single" => OutputStyle.Single
      case "per_bank" | "per_segment" => OutputStyle.PerBank
      case "lunix" => OutputStyle.LUnix
      case x => log.fatal(s"Invalid output style: `$x`")
    }
    if (outputStyle != OutputStyle.PerBank && outputPackagers.nonEmpty) {
      log.error("Different output formats per segment are allowed only if style=per_segment")
    }
    val debugOutputFormatName = os.get(classOf[String], "labels", "vice")
    val debugOutputFormat = DebugOutputFormat.map.getOrElse(
      debugOutputFormatName.toLowerCase(Locale.ROOT),
      log.fatal(s"Invalid label file format: `$debugOutputFormatName`"))

    val builtInFeatures = builtInCpuFeatures(cpu) ++ Map(
      "ENCODING_NOLOWER" -> toLong(codec.supportsLowercase),
      "ENCODING_SAME" -> toLong(codec.name == srcCodec.name),
      "DECIMALS_SAME" -> toLong(codec.stringTerminator == srcCodec.stringTerminator && (0 to 9).forall{c =>
        codec.encodeDigit(c) == srcCodec.encodeDigit(c)
      }),
      "ENCCONV_SUPPORTED" -> toLong((codec.name, srcCodec.name) match {
        // TODO: don't rely on names!
        case ("PETSCII", "CBM-Screen") |
             ("PETSCII-JP", "CBM-Screen-JP") |
             ("ATASCII", "ATASCII-Screen") =>
          CpuFamily.forType(cpu) == CpuFamily.M6502
        case ("Color-Computer", "Color-Computer-Screen") =>
          CpuFamily.forType(cpu) == CpuFamily.M6809
        case _ => codec.name == srcCodec.name
      }),
      "NULLCHAR_SAME" -> toLong(codec.stringTerminator == srcCodec.stringTerminator)
    )

    import scala.collection.JavaConverters._
    val ds = conf.getSection("define")
    val definedFeatures = ds.getKeys().asScala.toList.map { k =>
      val value = ds.get(classOf[String], k).trim() match {
        case "true" | "on" | "yes" => 1L
        case "false" | "off" | "no" | "" => 0L
        case x => x.toLong
      }
      k -> value
    }.toMap

    var actualStartingModules = startingModules
    if (ramInitialValuesBank.isDefined) {
      actualStartingModules = "init_rw_memory" :: actualStartingModules
    }

    new Platform(
      cpu,
      flagOverrides,
      actualStartingModules,
      codec,
      srcCodec,
      builtInFeatures ++ definedFeatures,
      outputPackagers,
      defaultOutputPackager,
      codeAllocators.toMap,
      variableAllocators.toMap,
      zpRegisterSize,
      freeZpBytes,
      if (fileExtension == "" || fileExtension.startsWith(".")) fileExtension else "." + fileExtension,
      generateBbcMicroInfFile,
      generateGameBoyChecksums,
      bankNumbers,
      bankLayouts,
      bankFills,
      defaultCodeBank,
      ramInitialValuesBank,
      debugOutputFormat,
      outputStyle)
  }

  @inline
  private def toLong(b: Boolean): Long = if (b) 1L else 0L

  def builtInCpuFeatures(cpu: Cpu.Value): Map[String, Long] = {
    Map[String, Long](
      "ARCH_6502" -> toLong(CpuFamily.forType(cpu) == CpuFamily.M6502),
      "CPU_6502" -> toLong(Set(Cpu.Mos, Cpu.StrictMos, Cpu.Ricoh, Cpu.StrictRicoh)(cpu)),
      "CPU_65C02" -> toLong(cpu == Cpu.Cmos),
      "CPU_65CE02" -> toLong(cpu == Cpu.CE02),
      "CPU_65816" -> toLong(cpu == Cpu.Sixteen),
      "CPU_HUC6280" -> toLong(cpu == Cpu.HuC6280),
      "ARCH_I80" -> toLong(CpuFamily.forType(cpu) == CpuFamily.I80),
      "CPU_Z80" -> toLong(cpu == Cpu.Z80),
      "CPU_EZ80" -> toLong(cpu == Cpu.EZ80),
      "CPU_8080" -> toLong(cpu == Cpu.Intel8080),
      "CPU_8085" -> toLong(cpu == Cpu.Intel8085),
      "CPU_GAMEBOY" -> toLong(cpu == Cpu.Sharp),
      "ARCH_X86" -> toLong(CpuFamily.forType(cpu) == CpuFamily.I86),
      "CPU_8086" -> toLong(cpu == Cpu.Intel8086),
      "CPU_80186" -> toLong(cpu == Cpu.Intel80186),
      "ARCH_6800" -> toLong(CpuFamily.forType(cpu) == CpuFamily.M6800),
      "ARCH_6809" -> toLong(CpuFamily.forType(cpu) == CpuFamily.M6809),
      "ARCH_ARM" -> toLong(CpuFamily.forType(cpu) == CpuFamily.ARM),
      "ARCH_68K" -> toLong(CpuFamily.forType(cpu) == CpuFamily.M68K)
      // TODO
    )
  }

  def parseNumberOrRange(s:String, step: Int)(implicit log: Logger): Seq[Int] = {
    if (s.contains("-")) {
      val segments = s.split("-")
      if (segments.length != 2) {
        log.fatal(s"Invalid range: `$s`")
      }
      Range(parseNumber(segments(0).trim()), parseNumber(segments(1).trim()) + 1, step)
    } else {
      Seq(parseNumber(s.trim()))
    }
  }

  def parseNumber(str: String): Int = {
    val s = str.trim
    if (s.startsWith("$")) {
      Integer.parseInt(s.substring(1), 16)
    } else if (s.startsWith("0x")) {
      Integer.parseInt(s.substring(2), 16)
    } else if (s.startsWith("0X")) {
      Integer.parseInt(s.substring(2), 16)
    } else if (s.startsWith("%")) {
      Integer.parseInt(s.substring(1), 2)
    } else if (s.startsWith("0b")) {
      Integer.parseInt(s.substring(2), 2)
    } else if (s.startsWith("0B")) {
      Integer.parseInt(s.substring(2), 2)
    } else if (s.startsWith("0o")) {
      Integer.parseInt(s.substring(2), 8)
    } else if (s.startsWith("0O")) {
      Integer.parseInt(s.substring(2), 8)
    } else if (s.startsWith("0q")) {
      Integer.parseInt(s.substring(2), 4)
    } else if (s.startsWith("0Q")) {
      Integer.parseInt(s.substring(2), 4)
    } else {
      s.toInt
    }
  }

  def parseSymbolAndBonus(token: String): (String, Int) = {
    if (token.contains("+")) {
      token.split("\\+", 2) match {
        case Array(n, b) => n.trim -> parseNumber(b)
      }
    } else if (token.contains("-")) {
      token.split("-", 2) match {
        case Array(n, b) => n.trim -> -parseNumber(b)
      }
    } else {
      token.trim -> 0
    }
  }
}